documentation/api/metrics.md

```diff
@@ -0,0 +1,164 @@
+# HELP systemruntime_threadpool_queue_length ThreadPool Queue Length
+# TYPE systemruntime_threadpool_queue_length gauge
+systemruntime_threadpool_queue_length 0 1618889166338
+systemruntime_threadpool_queue_length 0 1618889176315
+systemruntime_threadpool_queue_length 0 1618889186313
+# HELP systemruntime_threadpool_completed_items_count ThreadPool Completed Work Item Count
+# TYPE systemruntime_threadpool_completed_items_count gauge
+systemruntime_threadpool_completed_items_count 2 1618889166338
+systemruntime_threadpool_completed_items_count 2 1618889176315
+systemruntime_threadpool_completed_items_count 2 1618889186313
+# HELP systemruntime_time_in_gc_ratio % Time in GC since last GC
+# TYPE systemruntime_time_in_gc_ratio gauge
+systemruntime_time_in_gc_ratio 0 1618889166338
+systemruntime_time_in_gc_ratio 0 1618889176315
+systemruntime_time_in_gc_ratio 0 1618889186313
+# HELP systemruntime_gen_0_size_bytes Gen 0 Size
+# TYPE systemruntime_gen_0_size_bytes gauge
+systemruntime_gen_0_size_bytes 0 1618889166338
+systemruntime_gen_0_size_bytes 0 1618889176315
+systemruntime_gen_0_size_bytes 0 1618889186313
+# HELP systemruntime_gen_1_size_bytes Gen 1 Size
+# TYPE systemruntime_gen_1_size_bytes gauge
+systemruntime_gen_1_size_bytes 0 1618889166338
+systemruntime_gen_1_size_bytes 0 1618889176316
+systemruntime_gen_1_size_bytes 0 1618889186313
+# HELP systemruntime_gen_2_size_bytes Gen 2 Size
+# TYPE systemruntime_gen_2_size_bytes gauge
+systemruntime_gen_2_size_bytes 0 1618889166339
+systemruntime_gen_2_size_bytes 0 1618889176316
+systemruntime_gen_2_size_bytes 0 1618889186313
+# HELP systemruntime_loh_size_bytes LOH Size
+# TYPE systemruntime_loh_size_bytes gauge
+systemruntime_loh_size_bytes 0 1618889166339
+systemruntime_loh_size_bytes 0 1618889176316
+systemruntime_loh_size_bytes 0 1618889186313
+# HELP systemruntime_alloc_rate_bytes Allocation Rate
+# TYPE systemruntime_alloc_rate_bytes gauge
+systemruntime_alloc_rate_bytes 40900 1618889166339
+systemruntime_alloc_rate_bytes 40900 1618889176316
+systemruntime_alloc_rate_bytes 40900 1618889186313
+# HELP systemruntime_assembly_count Number of Assemblies Loaded
+# TYPE systemruntime_assembly_count gauge
+systemruntime_assembly_count 40 1618889166339
+systemruntime_assembly_count 40 1618889176316
+systemruntime_assembly_count 40 1618889186313
+# HELP systemruntime_active_timer_count Number of Active Timers
+# TYPE systemruntime_active_timer_count gauge
+systemruntime_active_timer_count 0 1618889166339
+systemruntime_active_timer_count 0 1618889176316
+systemruntime_active_timer_count 0 1618889186313
+```
+
+## Supported Runtimes
+
+| Operating System | Runtime Version |
+|---|---|
+| Windows | .NET Core 3.1, .NET 5+ |
+| Linux | .NET Core 3.1, .NET 5+ |
+| MacOS | .NET Core 3.1, .NET 5+ |
+
+## Additional Notes
+
+### Custom Metrics
+
+The metrics providers and counter names to return from this route can be specified via configuration. See [TBD]() for more information.
```

Comment:
## Additional Configuration

### Metrics Urls

In addition to the ordinary diagnostics urls that `dotnet-monitor` binds to, it also binds to metric urls that only expose the `/metrics` endpoint. Unlike the other endpoints, the metrics urls do not require authentication. Unless you enable collection of custom providers that may contain sensitive business logic, it is generally consider safe to expose metrics endpoints. 

Metrics urls can configured via the command line:

```cmd
dotnet monitor collect --metricUrls http://*:52325/
```

Or configured via configuration:

```json
{
  "Metrics": {
    "Endpoints": "http://localhost:52325"
  }
}
```

### Customize collection interval and counts

In the default configuration, `dotnet-monitor` requests that the connected runtime provides updated counter values every 10 seconds and will retain 3 data point for every collected metric. When using a collection tool like Prometheus, it is recommended that you set your scrape interval to `MetricCount` * `UpdateIntervalSeconds`. In the default configuration, we recommend you scrape `dotnet-monitor` for metrics every 30 seconds.

You can customize the number of data points stored per metric and the frequency at which the runtime updates each metric via the following configuration:

```json
{
  "Metrics": {
    "MetricCount": 3,
    "UpdateIntervalSeconds": 10,
  }
}
```

### Custom Metrics

Additional metrics providers and counter names to return from this route can be specified via configuration. 

```json
{
  "Metrics": {
    "Providers": [
      {
        "ProviderName": "Microsoft-AspNetCore-Server-Kestrel",
        "CounterNames": [
          "connections-per-second",
          "total-connections"
        ]
      }
    ]
  }
}
```

#### Allow insecure channel for custom metrics

In the default configuration, enabling of custom metrics changes the default binding address from `http` to `https` for the metrics urls. To continue egressing custom metric via an endpoint with `http` scheme, you explicit opt-in to it. You can do so via the following configuration:

```json
{
  "Metrics": {
    "AllowInsecureChannelForCustomMetrics": true
  }
}
```

### Disable default providers

In addition to enabling custom providers, `dotnet-monitor` also allows you to disable to collection of the default providers. You can do so via the following configuration:

```json
{
  "Metrics": {
    "IncludeDefaultProviders": false
  }
}

Reply:
I think all of this should go into the configuration doc you wrote and that the HTTP API really should only document the HTTP aspects. I have a TBD link to point to the metrics section when you merge the configuration PR.

Reply:
Agreed